src/coreclr/tools/Common/TypeSystem/IL/Stubs/AsyncThunks.cs

```diff
@@ -1,37 +1,231 @@
+
+                    MethodDesc fromExceptionMd;
+                    if (logicalReturnType != null)
+                    {
+                        MethodSignature fromExceptionSignature = new MethodSignature(
+                            MethodSignatureFlags.Static,
+                            genericParameterCount: 1,
+                            returnType: ((MetadataType)returnType.GetTypeDefinition()).MakeInstantiatedType(context.GetSignatureVariable(0, true)),
+                            parameters: new[] { exceptionType });
+
+                        fromExceptionMd = context.SystemModule
+                            .GetKnownType("System.Runtime.CompilerServices"u8, "AsyncHelpers"u8)
+                            .GetKnownMethod(isValueTask ? "ValueTaskFromException"u8 : "TaskFromException"u8, fromExceptionSignature)
+                            .MakeInstantiatedMethod(new Instantiation(logicalReturnType));
+                    }
+                    else
+                    {
+                        MethodSignature fromExceptionSignature = new MethodSignature(
+                            MethodSignatureFlags.Static,
+                            genericParameterCount: 0,
+                            returnType: returnType,
+                            parameters: new[] { exceptionType });
+
+                        fromExceptionMd = context.SystemModule
+                            .GetKnownType("System.Runtime.CompilerServices"u8, "AsyncHelpers"u8)
+                            .GetKnownMethod(isValueTask ? "ValueTaskFromException"u8 : "TaskFromException"u8, fromExceptionSignature);
+                    }
+
+                    codestream.Emit(ILOpcode.call, emitter.NewToken(fromExceptionMd));
+                    codestream.EmitStLoc(returnTaskLocal);
+                    codestream.Emit(ILOpcode.leave, returnTaskLabel);
+                    codestream.EndHandler(tryCatchRegion);
+                }
+
+                codestream.EmitLabel(suspendedLabel);
+
+                MethodDesc finalizeTaskReturningThunkMd;
+
+                if (isValueTask)
+                {
+                    finalizeTaskReturningThunkMd = context.SystemModule
+                        .GetKnownType("System.Runtime.CompilerServices"u8, "AsyncHelpers"u8)
+                        .GetKnownMethod("FinalizeValueTaskReturningThunk"u8, null);
+                }
+                else
+                {
+                    finalizeTaskReturningThunkMd = context.SystemModule
+                        .GetKnownType("System.Runtime.CompilerServices"u8, "AsyncHelpers"u8)
+                        .GetKnownMethod("FinalizeTaskReturningThunk"u8, null);
+                }
+
+                if (logicalReturnType != null)
+                {
+                    finalizeTaskReturningThunkMd = finalizeTaskReturningThunkMd.MakeInstantiatedMethod(new Instantiation(logicalReturnType));
+                }
```

Comment:
This has the same bug as https://github.com/dotnet/runtime/pull/121435#discussion_r2513125381. There are not just two, there are 4 methods we need to choose from. Each of these has two overloads.